backend/data_preparation/extractor/fire_extractor.py

```diff
@@ -0,0 +1,175 @@
+                    len(shp.record(0)['PERDATTIME']) > 11 else datetime.datetime.strptime(shp.record(0)['PERDATTIME'],
+                                                                                          '%m/%d/%Y')
+                # For some records after 2016, record schema for firename is: FIRENAME, for agency is: AGENCY, for
+                # datetime is: PERDATTIME
+            try:
+                result["area"] = float(shp.record(0)["GISACRES"])
+                # after 2016, record schema for area is: GISACRES
+            except IndexError:
+                result["area"] = float(shp.record(0)["gisAcres"])
+                # For some records after 2016, record schema for area is: gisAcres
+        geom = self.extract_full_geom(shp)
+        # basic geom is the full geom of this shapefile record
+        # see function extract_full_geom below
+        result["geopolygon_full"] = str(geom)
+        # geopolygon_full is the full geom of this shapefile record
+        result["geopolygon_large"] = str(self.simplify_multipolygon(geom,1.e-04))
+        # geopolygon_large is the simpilfied full geom of this shapefile record, threshold is 1.e-04
+        result["geopolygon_medium"] = str(self.simplify_multipolygon(geom,1.e-03))
+        # geopolygon_medium is the simpilfied full geom of this shapefile record, threshold is 1.e-03
+        result["geopolygon_small"] = str(self.simplify_multipolygon(geom,1.e-02))
+        # geopolygon_small is the simpilfied full geom of this shapefile record, threshold is 1.e-02
+        result["year"] = year
+        # attribute "year" is the year extracted from record file names
+        result["if_sequence"] = if_sequence
+        # if_sequence is passed as a parameter. If in the temp folder, there are more than one set of records, then
+        # if_sequence is True
+        result["id"] = id
+        # id is passed as a parameter
+        result["state"] = state
+        # state is passed as a parameter
+        return result
+
+    @staticmethod
+    def extract_full_geom(shp: shapefile.Reader) -> MultiPolygon:
+        """
+        Extracts a full geom from a shp reader
+        :param shp: shapefile.Reader
+        :return: Multipolygon
+        """
+        fire = shp.shapeRecord(0).shape.__geo_interface__
+        geom = shape(fire)
+        return geom
+
+    @staticmethod
+    def simplify_multipolygon(multipolygon: MultiPolygon, threshold: float) -> MultiPolygon:
+        """
+        Simplifies all components of a multipolygon
+        :param multipolygon:shapely.geometry.Multipolygon or shapely.geometry.polygon.Polygon
+        :param threshold:float, the threshold of simplification
+        :return:shapely.geometry.Multipolygon
+        """
+        try:
+            # if polygon is a shapely.geometry.MultiPolygon, then it can be convert into a list of
+            # shapely.geometry.Polygon
+            polygons = list(multipolygon)
+        except TypeError:
```

Comment:
when would typeError got raised? multi polygon cannot be listed? (that is, it is not an iterable?)